Fix crash during alarm state evaluation if attribute value is not set (#330)

mliszcz opened the pull request:
Fix for following scenario:
* there is an alarm configured for an attribute
* Device::dev_state is evaluated
* attribute's read callback throws DevFailed before calling set_value
* attempt to evaluate alarm fails (segfault during access to value which is not set)

Proposed correction is to silently skip such attribute during alarm evaluation (it will not contribute to device's ALARM state). @bourtemb, is this acceptable solution?

I have two other points I'd like to discuss:
* To simplify the reproduction, I had to reset attribute's value pointer after it was deleted. Otherwise something else was allocated in that memory and I was getting garbage read instead of a segfault. I propose to define an action to reset all the pointers after corresponding memory is deleted (applies only to pointers that can be reused, e.g. global or member variables). We may find some new issues that way.
* There is a lot of duplicated code in `attribute.cpp` and in `attrsetval.cpp` for different data types. It could be easily removed and replaced with template code. Maybe we should define an action to remove duplications (for now I've re-set only `value.lg_seq` - see previous point).

BTW all commits are failed on appveyor because MSVC don't like `not` operator. I've fixed this just now.

mliszcz commented:
@t-b thanks for checking this PR.

> Setting the pointer to nullptr after deleting sounds like something we want to do for the other types as well. What is the reason why that is only done for this type?

Zeroing the pointer is not necessary for the correction to work. It is just to simplify the reproduction.

I agree it is a good thing to reset the pointers, but I'd like to have a dedicated issue to fix this kind of problem in the whole codebase. Some examples in other places:

https://github.com/tango-controls/cppTango/blob/ab05abc289d187130614966e4c79e5390ca41b16/cppapi/server/device.cpp#L2944-L2950

https://github.com/tango-controls/cppTango/blob/ab05abc289d187130614966e4c79e5390ca41b16/cppapi/server/device.h#L3683-L3684

https://github.com/tango-controls/cppTango/blob/ab05abc289d187130614966e4c79e5390ca41b16/cppapi/server/device.cpp#L5333-L5336

If the kernel team agrees that we need this kind of change, and if it is in [Tango 9 LTS project](https://github.com/tango-controls/cppTango/projects/5), I can work on this task and priotitize it over other S2Innovation activities.

Note that some refactoring will be needed because sometimes pointer to attribute data is passed by value and delete is done using copied pointer, not the original one.

CC @bourtemb

t-b commented:
@mliszcz 
> I agree it is a good thing to reset the pointers, but I'd like to have a dedicated issue to fix this kind of problem in the whole codebase. 

I think this would be a reasonable thing to do. We should then also define some templates which will do delete and set to null in one go. Having two templates would solve the issue of `delete` vs `delete []`.

bourtemb commented:
> > I agree it is a good thing to reset the pointers, but I'd like to have a dedicated issue to fix this kind of problem in the whole codebase.
> 
> I think this would be a reasonable thing to do. We should then also define some templates which will do delete and set to null in one go. Having two templates would solve the issue of `delete` vs `delete []`.

I agree this would be a reasonable thing to do. This will probably help to discover other issues.

andygotz commented:
Fixing all places in the library to reset pointers sounds like a big job potentially. How much work do you estimate this to be and how can we test all the changes? Do we have tests for all cases?

bourtemb commented:
> I agree it is a good thing to reset the pointers, but I'd like to have a dedicated issue to fix this kind of problem in the whole codebase. Some examples in other places:
> 
> https://github.com/tango-controls/cppTango/blob/ab05abc289d187130614966e4c79e5390ca41b16/cppapi/server/device.cpp#L2944-L2950

Well, in this specific case, ptr gets out of scope immediately after. So, in this specific case, it looks like there is no interest in resetting it.

I think we should do this only for pointers which can be reused somewhere else.

mliszcz commented:
> Fixing all places in the library to reset pointers sounds like a big job potentially. How much work do you estimate this to be and how can we test all the changes? Do we have tests for all cases?

From what I have checked, I think the problem is only with pointer(s) to attribute data. These pointers are deleted (at least) in: attribute.cpp, device.cpp, attrsetval.cpp in multiple places. 

Zeroing the pointer is quite trivial job, but the problem is sometimes there is no access to the original pointer, but to a copy. Thus, I estimate the effort to 16h (two days, for potential refactoring).

It is not possible to test such change.

> Well, in this specific case, ptr gets out of scope immediately after. So, in this specific case, it looks like there is no interest in resetting it.

> I think we should do this only for pointers which can be reused somewhere else.

This specific case also needs to be fixed.The *copy* of a pointer goes out of scope. See what is returned from `get_short_value`:
https://github.com/tango-controls/cppTango/blob/ab05abc289d187130614966e4c79e5390ca41b16/cppapi/server/attribute.h#L2186

```cpp
 case Tango::DEV_ENUM : 
 { 
     Tango::DevVarShortArray *ptr = att.get_short_value(); 
     a <<= *ptr; 
     delete ptr; 
     // somehow we need to reset: att.value.sh_seq = nullptr
     break; 
}
```

bourtemb commented:
> // somehow we need to reset: att.value.sh_seq = nullptr
@mliszcz, OK, I see your point now. Thanks for the clarification.

mliszcz commented:
@t-b thanks for the review! I've applied changes per your comments:
> fbcbc8a (Reset Attribute::value.lg_seq pointer after delete, 2019-05-26)

> Setting the pointer to nullptr after deleting sounds like something we want to do for the other types as well. What is the reason why that is only done for this type?

Changed to: Reset Attribute::value pointers in Attribute::delete_seq bf8ce6c
Pointers of other types are reset as well, in `del_seq` function. For global reset action there is #565.

> 49ad354 (Fix appveyor builds, 2019-05-27)
That should probably be folded into 0ef70cc as that commit introduced the incompatibility.

Also, this one was squashed into previous commit:
> 65e8e92 (Initialize exception flag for Long_attr in DevTest, 2019-05-26)

@bourtemb, @andygotz I've opened #565 to reset the pointers. Lets use it for further discussions on that topic and leave this PR for crash correction review.

bourtemb commented:
> Proposed correction is to silently skip such attribute during alarm evaluation (it will not contribute to device's ALARM state). @bourtemb, is this acceptable solution?

Very good question.
I talked a bit with some colleagues and we think it would be more consistent to switch the global state to UNKNOWN (if State evaluation was triggered (State was ON before evaluation)) in this specific case and to add an explanation in the status, with the received exception description...
It would be good to get some feedback from the users on this point actually.

>     * There is a lot of duplicated code in `attribute.cpp` and in `attrsetval.cpp` for different data types. It could be easily removed and replaced with template code. Maybe we should define an action to remove duplications (for now I've re-set only `value.lg_seq` - see previous point).

Yes, we can create an issue for that.

mliszcz commented:
Thanks @bourtemb!

> Very good question.
> I talked a bit with some colleagues and we think it would be more consistent to switch the global state to UNKNOWN (if State evaluation was triggered (State was ON before evaluation)) in this specific case and to add an explanation in the status, with the received exception description...
> It would be good to get some feedback from the users on this point actually.

I can change it the way you propose. Would it be possible for you to ask some users (I guess there is a lot of them at ESRF) for feedback?

> > ```
> > * There is a lot of duplicated code [...]
> > ```
> 
> Yes, we can create an issue for that.

Done: https://github.com/tango-controls/cppTango/issues/569

mliszcz commented:
Hi @bourtemb,
Thanks for triggering the discussion on the kernel list.
So far we have a few contradictory opinions. Do you think we can make any conclusion now?

Also, up until now we were discussing "exception" scenario (as in original issue), but the same problem will be visible if the user will simply forget to call `Attribute::set_value(...)`. i.e. all these shall crash equally:
```cpp
void DevTest::read_Long_attr(Tango::Attribute &att)
{
    Tango::Except::throw_exception(/* ... */);
    att.set_value(&attr_long);
}
void DevTest::read_Long_attr(Tango::Attribute &att)
{
    // do nothing
}
void DevTest::read_Long_attr(Tango::Attribute &att)
{
    att.set_quality(ATTR_CHANGING);
}
```

bourtemb commented:
> Hi @bourtemb,
> Thanks for triggering the discussion on the kernel list.
> So far we have a few contradictory opinions. Do you think we can make any conclusion now?

Indeed with have contradictory opinions.
I triggered the discussion with my colleagues at the ESRF and their opinion was that in the case of an exception thrown by an attribute with an alarm threshold defined, the device state should switch to ALARM and the status should report that this attribute has thrown an exception.

> 
> Also, up until now we were discussing "exception" scenario (as in original issue), but the same problem will be visible if the user will simply forget to call `Attribute::set_value(...)`. i.e. all these shall crash equally:
> 
> ```c++
> void DevTest::read_Long_attr(Tango::Attribute &att)
> {
>     Tango::Except::throw_exception(/* ... */);
>     att.set_value(&attr_long);
> }
> void DevTest::read_Long_attr(Tango::Attribute &att)
> {
>     // do nothing
> }
> void DevTest::read_Long_attr(Tango::Attribute &att)
> {
>     att.set_quality(ATTR_CHANGING);
> }
> ```
> 
> Do you think we need to differentiate between all these scenarios?

The use case where the user does not set the attribute value is actually currently considered as a bug on the device server programmer side. I think it is a mistake of the programmer...
Maybe we can prevent the device server to crash in this case...?

mliszcz commented:
Thanks!

> The use case where the user does not set the attribute value is actually currently considered as a bug on the device server programmer side. I think it is a mistake of the programmer...
Maybe we can prevent the device server to crash in this case...?

Crash will not happen once this PR is applied even in it's current state.

> the device state should switch to ALARM and the status should report that this attribute has thrown an exception.

I'll then update the PR accordingly. I propose the following:

* If exception is thrown during read:
  * State will be set to ALARM
    * `Attribute::check_alarm()` will return `true` if `get_value_flag()` is false
  * Status will be updated with exception's message
    * **I need to store this message/exception in `Attribute` class**
    * `MultiAttribute::read_alarm` will update status with the message if attribute `is_alarmed()` and `get_value_flag()` is false
* If set_value was never called:
  * State will be set to ALARM
    * `Attribute::check_alarm()` will return `true` if `get_value_flag()` is false
  * Status will be updated with sth like.: "No value set for attribute xxx"
    * `MultiAttribute::read_alarm` will update status if attribute `is_alarmed()` and `get_value_flag()` is false

Is this ok?

bourtemb commented:
I think we'll need to discuss a bit more and agree on this during the next Tango kernel teleconf meeting because there were some very different opinions on the kernel mailing list, and even among my colleagues, some of them had different opinions...
Be careful with the INVALID quality factor case where the value is not set and where I think the device state should not be switched to ALARM...

mliszcz commented:
Ok, I'll wait with any changes until we all agree on a way forward.

bourtemb commented:
> Proposed correction is to silently skip such attribute during alarm evaluation (it will not contribute to device's ALARM state). @bourtemb, is this acceptable solution?

It was agreed during today's Tango Kernel teleconf meeting to add a warning log when this use case occurs. 
State and Status will not be modified automatically by the Tango library in this use case.

mliszcz commented:
I've added a log:
```cpp
DEV_WARN_STREAM(dev) << "Attribute has no value, skipping alarm evaluation for: " << get_name() << std::endl;
```

mliszcz commented:
Thanks @bourtemb! I confirm there is an issue in current solution.
If alarm was already reported for a given attribute (which later starts to throw), we don't even run the alarm check again, but instead just check if quality has been changed and assume it is still in alarm state.

https://github.com/tango-controls/cppTango/blob/6a2857c32baba5454d936815279314e60b838b9b/cppapi/server/multiattribute.cpp#L1330-L1336

I'll correct this and update the PR.

mliszcz commented:
@bourtemb 
I've updated the test case to cover this scenario and moved the check to earlier step, immediately after exception is thrown from read callback: ca4bac405b72b34575d61e7a08733df635535bf3. I've set quality to INVALID, because (judging from the check few lines above), it is the only reasonable quality if attribute has no value. This was sufficient to skip alarm evaluation in `Multiattribute::check_alarm`.

```diff
diff --git a/cppapi/server/device.cpp b/cppapi/server/device.cpp
index ef21033b..3d28cb1f 100644
--- a/cppapi/server/device.cpp
+++ b/cppapi/server/device.cpp
@@ -1256,40 +1256,48 @@ Tango::DevState DeviceImpl::dev_state()
                                 continue;
                             }
                             attr_vect[att.get_attr_idx()]->read(this, att);
                             Tango::AttrQuality qua = att.get_quality();
                             if ((qua != Tango::ATTR_INVALID) && (att.get_value_flag() == false))
                             {
                                 TangoSys_OMemStream o;

                                 o << "Read value for attribute ";
                                 o << att.get_name();
                                 o << " has not been updated";
                                 o << "Hint: Did the server follow Tango V5 attribute reading framework ?" << std::ends;

                                 Except::throw_exception((const char *) API_AttrValueNotSet, o.str(),
                                                         (const char *) "DeviceImpl::dev_state");
                             }
                         }
                     }
                     catch (Tango::DevFailed &)
                     {
+                        if (!att.get_value_flag())
+                        {
+                            WARN_STREAM
+                                << "Attribute has no value, forcing INVALID quality for: "
+                                << att.get_name() << std::endl;
+                            att.set_quality(Tango::ATTR_INVALID);
+                        }
+
```
